angr/knowledge_plugins/sync/sync_controller.py

```diff
@@ -0,0 +1,376 @@
+        #func_addr = func_addr if func_addr else -1
+        sync_cmt = binsync.data.Comment(func_addr, addr, comment, decompiled)
+
+        return state.set_comment(sync_cmt)
+
+    @init_checker
+    @make_state
+    @last_push
+    def push_comments(self, comments: List[binsync.data.Comment], user=None, state=None):
+        """
+        Push a bunch of comments upwards.
+
+        :param list comments:   A list of BinSync Comments
+        :return:                bool
+        """
+
+        r = True
+        for cmt in comments:
+            r &= state.set_comment(cmt)
+        return r
+
+    @init_checker
+    @make_state
+    @last_push
+    def push_stack_variables(self, stack_variables: List[SimStackVariable], var_manager: VariableManagerInternal,
+                             user=None, state=None):
+        """
+
+        :param stack_variables:
+        :param var_manager:
+        :return:
+        """
+        r = True
+        for var in stack_variables:
+            guessed_var_type = var_manager.get_variable_type(var)
+            var_type = guessed_var_type if guessed_var_type else "BOT"
+
+            # construct a StackVariable for each SimStackVariable
+            sync_stack_var = StackVariable(var.offset, StackOffsetType.ANGR, var.name,
+                                           var_type, var.size, var_manager.func_addr)
+
+            r &= state.set_stack_variable(var_manager.func_addr, var.offset, sync_stack_var)
+
+        # return true only if all pushed worked
+        return r
+
+    @init_checker
+    @make_state
+    @last_push
+    def push_stack_variable(self, func_addr, offset, name, type_, size_, user=None, state=None):
+        sync_var = StackVariable(offset, StackOffsetType.ANGR, name, type_, size_, func_addr)
+        return state.set_stack_variable(func_addr, offset, sync_var)
+
+    #
+    # Pullers
+    #
+
+    @init_checker
+    @make_ro_state
+    def pull_function(self, addr, user=None, state=None) -> binsync.data.Function:
+        """
+        Pull a function downwards.
+
+        :param int addr:    Address of the function.
+        :param str user:    Name of the user.
+        :return:            The binsync.data.Function object if pulling succeeds, or None if pulling fails.
+        :rtype:             binsync.data.Function
+        """
+
+        func_addr = self.get_func_addr_from_addr(addr)
+        func_addr = func_addr if func_addr else -1
+
+        try:
+            func = state.get_function(func_addr)
+            return func
+        except KeyError:
+            return None
+
+    @init_checker
+    @make_ro_state
+    def pull_comment(self, addr, user=None, state=None) -> binsync.data.Comment:
+        """
+        Pull a comment downwards.
+
+        :param int addr:    Address of the comment.
+        :param str user:    Name of the user.
+        :return:            The comment it self, or None if there is no comment.
+        :rtype:             str or None
```

Comment:
What's the return type? `str or None` is not `binsync.data.Comment`.

Reply:
good catch

Reply:
It's binsync.data.Comment or None